src/hnsw/htab_cache.c

```diff
@@ -0,0 +1,53 @@
+#include "htab_cache.h"
+
+HTABCache cache_create(char *name, MemoryContext ctx)
+{
+    HTABCache cache;
+
+    HASHCTL hctl;
+    hctl.keysize = sizeof(HTABCacheKey);
+    hctl.entrysize = sizeof(HTABCacheEntry);
+    hctl.hcxt = ctx;
+    HTAB *htab = hash_create(name, 1, &hctl, HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+    cache.hctl = hctl;
+    cache.htab = htab;
+    return cache;
+}
+
+bool cache_remove(HTABCache *cache, HTABCacheKey *key)
+{
+    bool status;
+
+    hash_search(cache->htab, key, HASH_REMOVE, &status);
+
+    return status;
+}
+
+void *cache_get_item(HTABCache *cache, HTABCacheKey *key)
+{
+    bool status;
+
+    HTABCacheEntry *item = (HTABCacheEntry *)hash_search(cache->htab, key, HASH_FIND, &status);
+
+    if(!status) {
+        return NULL;
+    }
+
+    return item->value;
+}
+
+void cache_set_item(HTABCache *cache, HTABCacheKey *key, void *value)
+{
+    HTABCacheEntry *entry;
+
+    entry = (HTABCacheEntry *)hash_search(cache->htab, key, HASH_ENTER, NULL);
+    entry->value = value;
+}
+
+void cache_destroy(HTABCache *cache)
+{
+    MemoryContext old_context = MemoryContextSwitchTo(cache->hctl.hcxt);
```

Comment:
Memory context ownership is a little strange here. 
It seems the memory context is constructed externally, by the caller, but it is destroyed internally, by this abstract data type. Is there a reason for this?

Reply:
I was thinking that if I create memory context with the same name they will share the same memory, but I looked through the source code and it seems name is mostly used on errors to display on which memory context the error was appeared. I think it is not critical for us to share the same name of MemoryContext among different cache instances, so I brought back memory context creation inside `cache_create`